### PR TITLE
Review, please -- "simple" example for command-line program

### PR DESCRIPTION
**Pull request by binkley**
For issue #1540

This PR is more for review & guidance than for merging (though it is mergeable best I can tell).
I'm looking to see if I'm heading in the right direction before I go too far along.

Things I'd like to understand/address include:
- How should I test this code?
- Is it OK to create an executable uberjar?  (I'm using this for manual command-line testing)
- Is there a better way to detect "isatty"?
- Did I pick the right directory and/or package for this work?

### CODE REVIEW

**Comment by binkley** [comment]
The initial commits are all code.

Before I write up separate documentation, I'd like working code to back up those docs.

**Comment by remkop** [comment]
Hi, sorry for the late response. I have been traveling with family. It may take me some time to get to this.

File: picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        }
+        // Use case #2 -- pipe or shell redirection
+        else if (null == System.console()) return runFromInput();
+        // Use case #3 -- interactive REPL
+        else return runRepl(prompt);
+    }
+
+    /** @todo Nicer way using more modern Java. */
+    private static Integer runFromInput() {
+        try (final BufferedReader lineReader = lineReader()) {
```
Comment (remkop) [inline picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java:73]: This method seems to be a rudimentary re-implementation of JLine functionality.
I don't think we should suggest to users that they do this.
My recommendation to users would be to familiarize themselves with what JLine offers and use that instead.

File: picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+final class Repl {
+    /**
+     * This is an <em>absolutely minimal</em> REPL.
+     *
+     * @param prompt the prompt to show on left-side asking user for input
+     *
+     * @return the exit code for <code>main()</code>
+     *
+     * @todo Newer versions of Java permit defining this as a lambda
```
Comment (remkop) [inline picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java:99]: Do we need this TODO?

File: picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    static Integer runRepl(final String prompt) {
+        try (final Terminal terminal = TerminalBuilder.builder().name("example tool").build()) {
+            final LineReader lineReader = LineReaderBuilder.builder().terminal(terminal).build();
+            do {
+                final String input;
+                try {
+                    input = lineReader.readLine(prompt);
+                } catch (final UserInterruptException | EndOfFileException e) {
+                    // Handling user banging on <kbd>Ctrl-C</kbd>
```
Comment (remkop) [inline picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java:109]: Do we want to catch these exceptions here?
My preference would be to let this method throw these exceptions and handle them in the place where this method is called.

File: picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                // TODO: It is challenging in Java to return _both_ a status
+                //  code for the command line, and to let caller handle
+                //  exceptions
+                final Integer result = process(input);
+                // Another policy decision: Should invalid input kill the
+                // REPL, or notify the user and continue?
+                if (0 != result) return result;
+            } while (true);
+        } catch (final IOException e) {
+            // TODO: It is challenging in Java to return _both_ a status code
```
Comment (remkop) [inline picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java:126]: If processing the exception is challenging them perhaps we should not catch it here, or perhaps we should wrap it in a custom exception and re-throw it .

File: picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * Your processing logic goes here. It could be human-edited text from the
+ * REPL, it could be tests, it could be text passed from the command line, it
+ * could be text read from <code>STDIN</code> via a pipe or shell
+ * redirection.
+ */
+final class YourProcessing {
+    // TODO: It is challenging in Java to return _all of_ a status code for
+    //  the command line, output results, and to let caller handle exceptions
```
Comment (remkop) [inline picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java:141]: Can we remove this TODO?

**Review by remkop — CHANGES_REQUESTED** [review]
Overall, this approach is not one I would take.

This approach uses picocli as an entry point, from which a JLine (or JLine-like) REPL is started, after which the `YourProcessing` class somehow needs to deal with the user input.

One problem of this approach is that it is not clear is how the application would tokenize the user input: given an input string of `mycommand --option --option2=value a b x,y,z` we want to break this up into `mycommand`, `--option`, `--option2=value`, `a`, `b`, and `x,y,z`, so we can invoke `CommandLine.execute(args)` with these tokens. JLine3 provides this functionality, and much more, out of the box. IMO it pays to spend the time to become familiar with the functionality that JLine3 offers rather than re-implementing this functionality ourselves.

The example in the https://github.com/remkop/picocli/blob/master/picocli-shell-jline3/README.md uses the JLine3 API extensively, and is more in line with what I would recommend to application authors. 

The Example in the README shows how to wire together picocli commands with JLine3.
Once that wiring is in place, your REPL application can support any number of commands, that can simply use the picocli API (`Runnable` or `Callable` classes with `@Option` or `@Parameters`-annotated fields) to define these commands.

There is a learning curve to this, but on the other hand it allows for application code to work at a much higher level of abstraction.

**Pull request closed without merging**